src/helpers.c

```diff
@@ -246,8 +246,15 @@ cr_repodata_blacklist(const char *repodata_path,
             continue;
         }
 
-        *blacklist = g_slist_prepend(*blacklist,
-                                     g_path_get_basename(rec->location_href));
+        if (g_strcmp0(rec->type, "modules") != 0) {
+            // Never blacklist the modules YAML
+            // This is a stopgap measure until createrepo_c learns to generate
+            // the modulemd.yaml itself. In the meantime, we just need to
+            // ensure that it never deletes the file.
+            *blacklist =
+                g_slist_prepend(*blacklist,
+                                g_path_get_basename(rec->location_href));
+        }
```

Comment:
I don't think this is a good idea. There are valid use-cases for being able to decompose a modular repo into regular rpm-md ones, or constructing a custom subset of modules in a repo.

Reply:
Can you be more specific about what you'd rather see? Right now, passing `--update` results in modules.yaml being deleted completely. This at least ensures that they aren't lost.

I'm not clear on what it should be doing instead.

Reply:
I think the point of this PR is to give `createrepo_c` awareness of what to do with `modules.yaml`. That means I would expect no logic like this to exist in the final form in _absolute_ and _unconditional_ terms. That said, if a switch was added for preserving `modules.yaml` (like we have for other auxiliary metadata), then I'd be okay with it.

Reply:
Unfortunately, I have to disagree, because one of the fundamental problems is that the file needs to be validated for components existing in the repo. If they don't exist, it needs to do something about it. It can't treat it as immutable or always correct either, because it's not the source of truth of the repository, createrepo_c is.

Reply:
@Conan-Kudo Would you mind spelling out what you see as the validation cases (and what should happen if there's a discrepancy)?

Here are the cases I can think of:

## Prerequisites
A modulemd YAML document is provided to createrepo_c that contains a mixture of zero or more
* Module Stream documents
* Module Default documents
* Module Translation documents

## Case 1: Entirely valid modulemd
Conditions:
* Module Stream documents: all entries in the `data.artifacts.rpms` section match RPM filenames in the repository being manipulated
* Module Default documents: all defaults reference a module for which there is at least one Module Stream document in the stream.
* Module Translation documents: all translations reference a module for which there is at least one Module Stream document in the stream.

This is the absolute easiest case. This modulemd YAML should be preserved in its entirety.

## Case 2: Some missing artifacts
Conditions:
* Module Stream documents: One or more of these documents have some entries in the `data.artifacts.rpms` section that are not reflected in the repository and some that are.
* Module Default documents: all defaults reference a module for which there is at least one Module Stream document in the stream.
* Module Translation documents: all translations reference a module for which there is at least one Module Stream document in the stream.

Do we filter out this module stream object or assume that as long as at least one item in the artifacts list is present that we should keep this stream document? I think the latter, since to do otherwise means that such artifacts would then be viewed by DNF as if they were non-modular and therefore considered for updates where they should not be.

## Case 3: All missing artifacts
Conditions:
* Module Stream documents: One or more of these documents have their `data.artifacts.rpms` section containing no artifacts that match the contents of the repository.
* Module Default documents: all defaults reference a module for which there is at least one Module Stream document in the stream.
* Module Translation documents: all translations reference a module for which there is at least one Module Stream document in the stream.

In this case, do we filter the offending Module Stream document(s) out of the final output?

## Case 4: Defaults with no matching stream
* Module Stream documents: all entries in the `data.artifacts.rpms` section match RPM filenames in the repository being manipulated
* Module Default documents: One or more such documents do not reference a module stream present in the metadata.
* Module Translation documents: all translations reference a module for which there is at least one Module Stream document in the stream.

In this situation, I think we still need to preserve the defaults, because this may be a metadata-only change (such as a case where the `fedora` repository contains a module stream and later we modify the stream defaults to treat that stream as the default. This doesn't require a rebuild of the module stream, so it may not be present in the updates repo.

## Case 5: Translations with no matching stream
* Module Stream documents: all entries in the `data.artifacts.rpms` section match RPM filenames in the repository being manipulated
* Module Default documents: all defaults reference a module for which there is at least one Module Stream document in the stream.
* Module Translation documents: one or more translations reference modules not represented in this YAML stream.

In this situation, I think we still need to preserve the translations, because this may be a metadata-only change (such as a case where the `fedora` repository contains a module stream and the only change to it is to add new translations. This doesn't require a rebuild of the module stream, so it may not be present in the updates repo.

Reply:
@sgallagher

Case 1 is fine. My concern with case 2 is that I don't know how it would be handled for modulemd aware reposyncs. In my view, if a component is missing, then it should be dropped from the metadata, because otherwise it'll look like the component is there when it isn't.

Cases 3, 4, and 5 should result in the module document getting dropped. What is the point of keeping a document that has no underlying content? It bloats the metadata for essentially no reason.

Reply:
> @sgallagher
> 
> Case 1 is fine. My concern with case 2 is that I don't know how it would be handled for modulemd aware reposyncs. In my view, if a component is missing, then it should be dropped from the metadata, because otherwise it'll look like the component is there when it isn't.

The problem with dropping the metadata in that situation is that we may end up with a repo that has modular RPMs that are treated by DNF as non-modular RPMs (meaning that stream filtering doesn't happen). This *will* cause issues.

Maybe we should just agree this is a hard failure for the createrepo operation and that if we end up with only a partial match, we assume that the repo was not properly populated. Reasonable?

> 
> Cases 3, 4, and 5 should result in the module document getting dropped. What is the point of keeping a document that has no underlying content? It bloats the metadata for essentially no reason.

Case 3, I agree with. If we have zero of the artifacts referenced for a stream, let's just drop that from the metadata. That will make life easier for filtering things out if we want to trim a repository.

For cases 4 and 5, I tried to explain in the original comment, but my view is that these are *valid* cases that we cannot reasonably be able to detect, short of duplicating data unnecessarily. It is completely reasonable that we might ship nodejs:10 as a module in Fedora 30's frozen `fedora` repository. Then, later in F30's life, we complete the default-modules-in-the-non-modular-buildroot feature (which this patch is a part of...) and we then push a metadata-only change to the modulemd that marks the nodejs:10 module as the default for Fedora 30, which takes over the update path from the non-modular RPM. If we require that the metadata must have a matching module stream, then it necessitates either rebuilding that module stream for the `fedora-updates` repo or else copying the data from the frozen `fedora` repo into the `fedora-updates` repo, wasting a lot of space. I'd much rather that we just always pass modulemd-defaults documents as-is. I think it's going to be pretty rare that this matters, but if we need to make a change like this and an overzealous filter gets in the way, it'll be difficult to track down and quite annoying.

Case 5 (translations), I think is even *more* likely to hit this than defaults, since new translations could arrive in the compose at any time, whereas defaults changes will be likely less common. We're working on automation to have pungi pull the latest translations from Zanata on each nightly compose and add the metadata to the modulemd document. If createrepo_c then chucks it back out, it will make that function largely useless.

Reply:
> Maybe we should just agree this is a hard failure for the createrepo operation and that if we end up with only a partial match, we assume that the repo was not properly populated. Reasonable?

That makes sense to me.

> For cases 4 and 5, I tried to explain in the original comment, but my view is that these are valid cases that we cannot reasonably be able to detect, short of duplicating data unnecessarily. It is completely reasonable that we might ship nodejs:10 as a module in Fedora 30's frozen fedora repository. Then, later in F30's life, we complete the default-modules-in-the-non-modular-buildroot feature (which this patch is a part of...) and we then push a metadata-only change to the modulemd that marks the nodejs:10 module as the default for Fedora 30, which takes over the update path from the non-modular RPM. If we require that the metadata must have a matching module stream, then it necessitates either rebuilding that module stream for the fedora-updates repo or else copying the data from the frozen fedora repo into the fedora-updates repo, wasting a lot of space. I'd much rather that we just always pass modulemd-defaults documents as-is. I think it's going to be pretty rare that this matters, but if we need to make a change like this and an overzealous filter gets in the way, it'll be difficult to track down and quite annoying.

Ugh, I see your point...

> Case 5 (translations), I think is even more likely to hit this than defaults, since new translations could arrive in the compose at any time, whereas defaults changes will be likely less common. We're working on automation to have pungi pull the latest translations from Zanata on each nightly compose and add the metadata to the modulemd document. If createrepo_c then chucks it back out, it will make that function largely useless.

I don't think translations for things that don't have any content associated with them would matter that much...

Reply:
@Conan-Kudo As this is now going into design of another feature, I'm moving this conversation to #134 

Please continue over there.